packages/react-native-web/src/exports/Image/index.js

```diff
@@ -94,56 +100,105 @@ function getFlatStyle(style, blurRadius, filterId) {
     }
-    const scaleSuffix = scale !== 1 ? `@${scale}x` : '';
-    uri = asset
-      ? `${asset.httpServerLocation}/${asset.name}${scaleSuffix}.${asset.type}`
-      : '';
-  } else if (typeof source === 'string') {
-    uri = source;
+
+    return resolveSource(source[0]);
   } else if (source && typeof source.uri === 'string') {
-    uri = source.uri;
+    resolvedSource = resolveObjectSource(source);
   }
 
-  if (uri) {
-    const match = uri.match(svgDataUriPattern);
-    // inline SVG markup may contain characters (e.g., #, ") that need to be escaped
+  if (resolvedSource.uri) {
+    const match = resolvedSource.uri.match(svgDataUriPattern);
     if (match) {
-      const [, prefix, svg] = match;
-      const encodedSvg = encodeURIComponent(svg);
-      return `${prefix}${encodedSvg}`;
+      resolvedSource = resolveSvgDataUriSource(resolvedSource, match);
     }
   }
 
-  return uri;
+  return resolvedSource;
+}
+
+// get the URI from the packager
+function resolveNumericSource(source: number): ImageSource {
+  const asset = getAssetByID(source);
+  if (asset == null) {
+    throw new Error(
+      `Image: asset with ID "${source}" could not be found. Please check the image source or packager.`
+    );
+  }
+  let scale = asset.scales[0];
+  if (asset.scales.length > 1) {
+    const preferredScale = PixelRatio.get();
+    // Get the scale which is closest to the preferred scale
+    scale = asset.scales.reduce((prev, curr) =>
+      Math.abs(curr - preferredScale) < Math.abs(prev - preferredScale)
+        ? curr
+        : prev
+    );
+  }
+
+  const scaleSuffix = scale !== 1 ? `@${scale}x` : '';
+  const uri = `${asset.httpServerLocation}/${asset.name}${scaleSuffix}.${asset.type}`;
+  const { height, width } = asset;
+
+  return { uri, height, width };
+}
+
+function resolveStringSource(source: string): ImageSource {
+  return { uri: source };
+}
+
+function resolveObjectSource(source: Object): ImageSource {
+  return (source: ImageSource);
+}
+
+function resolveSvgDataUriSource(
+  source: Object,
+  match: Array<string>
+): ImageSource {
+  const [, prefix, svg] = match;
+  // inline SVG markup may contain characters (e.g., #, ") that need to be escaped
+  const encodedSvg = encodeURIComponent(svg);
+
+  return {
+    ...source,
+    uri: `${prefix}${encodedSvg}`
+  };
+}
+
+// resolve any URI that might have a local blob URL created with `createObjectURL`
+function resolveBlobUri(source: ImageSource): string {
+  return ImageLoader.resolveBlobUri(source.uri);
+}
+
+function getSourceToDisplay(main: SourceState, fallback: SourceState) {
+  if (main.status === LOADED) return main.source;
+
+  // If there's no fallback URI, it's safe to use the main source URI
+  if (main.status === LOADING && !fallback.source.uri) {
+    // But it should not be used when the image would be loaded with custom headers
+    // Because the actual URI is only set (as a local blob url) after loading
+    if (!main.source.headers) return main.source;
```

Comment:
Can we list out all of the changes we are making in a list somewhere or draw attention to them with PR comments? I think a summary would help me review this better. It's hard to tell which things are related to the "headers" problem mentioned in the description and things like this.

Reply:
I've made sure both `source` and `defaultSource` can be loaded with headers

Both `source` and `defaultSource` are props implementing the `Source` interface, so both can be loaded with headers 
This is coming from existing code that performed similar checks about LOADING and defaultSource.uri